DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp

```diff
@@ -0,0 +1,123 @@
+#ifndef __CACHED_LOCAL_MEMORY_HPP__
+#define __CACHED_LOCAL_MEMORY_HPP__
+
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+
+#include "constexpr_math.hpp"   // DirectProgramming/DPC++FPGA/include
+
+namespace fpga_tools {
+
+template <typename T,             // type to store in the memory
+          size_t k_mem_depth,     // depth of the memory
+          size_t k_cache_depth    // number of elements in the cache
```

Comment:
Apologies if this is explained elsewhere, but will you suggest a better way (compared to the previous version) to determined the value of k_cache_depth? Or will it still be trial-and-error?

Reply:
I'm going to run some experiments on that.  If the method currently described in the readme really is required (ie if read latency of 1 has fmax implications that do not show up in the reports) then that is a clear reporting bug, and I'll file a case to get it fixed.  If it doesn't have fmax implications, then there's no point talking about it, and all you need to do is achieve II1, which is much simpler.

Reply:
Why can't the compiler figure this out? Realistically, you could create a cached RAM IP in RTL and this would happen transparently to the user (e.g. by applying an attribute such as `[[intel::memory_cached]]` to the array you want to have a cache on). The compiler can determine how deep to make the cache using static analysis at compile time.

Reply:
Paul what you're asking for is essentially for the compiler to modify the code at compile time.  It's not impossible, but also not a flow that's currently supported, and I don't think is likely to be supported.

Reply:
not at all: I'm asking for the compiler to use a RAM IP with a cache instead of whatever RAM IP we use by default. this won't affect the kernel behaviour at all since it's a local memory. This might be different with an off-chip memory because it would affect the side effects of the kernel to the outside world.

Reply:
Ah, I see what you're saying.  That's an excellent idea.  I was actually thinking about this as a way of working around the issues we discussed today related to ivdep.  I will be creating an internal case about the ivdep issues, I'll include the suggestion to have a cached RAM IP automatically used by the compiler, I think it will solve several issues and greatly simplify the user experience.